src/NLog/Targets/FileTarget.cs

```diff
@@ -2493,8 +2498,8 @@ private void WriteHeaderAndBom(BaseFileAppender appender)
             if (Header is null && !WriteBom) return;
 
             var length = appender.GetFileLength();
-            //  Write header and BOM only on empty files or if file info cannot be obtained.
-            if (length is null || length == 0)
+            //  Write header and BOM only on empty files or if file info cannot be obtained or WriteHeaderOnStartup is true.
+            if (length is null || length == 0 || WriteHeaderOnStartup)
```

Comment:
Please re-write the logic so it will not write additional BOM on non-empty-file, when enabling the new setting.

Reply:
I hope it's better now

Reply:
The logic is a little convoluted now, but hopefully the unit-test will ensure that logic continues to work.